### ISSUE TITLE
`PushCall` blocks indefinitely if remote side panics

### ISSUE THREAD

**Issue by Lachann** [creachadair/jrpc2#41]
During a `PushCall` call, if the other side panics or exits during the callback, `PushCall` will block forever without erroring. Setting a timeout on the `ctx` doesn't seem to help.

```
server:
	tctx, _ := context.WithTimeout(ctx, 1*time.Second)
	res, err := jrpc2.PushCall(tctx, "foo", "bar") // blocks forever
	if err != nil {
		fmt.Println(err)
		return "", err
	}
```

```
client:
	cli := jrpc2.NewClient(ch, &jrpc2.ClientOptions{
		OnCallback: func(ctx context.Context, req *jrpc2.Request) (interface{}, error) {
			panic("oh noes!")
			return "client", nil
		},
	})
```

**Comment by creachadair** [comment]
Can you say a bit more about the context where you observe this? Panicking out of the callback should terminate the client process, and thus terminate the channel to the server. Unless something swallows the panic (and without cleaning up the client), I can't see a good way to reproduce the symptom you describe here.

Re: Timeout. The server doesn't do anything with the context in a callback at present.

**Comment by creachadair** [comment]
I poked at this some more, and I think I see the issue: An execution error from the handler, as opposed to an error reported by the handler itself, is not propagated back to the server. That includes a panic, but will also include, for example, JSON marshaling problems.

Fortunately, the fix is straightforward, and I'll send a PR soon.

**Comment by creachadair** [comment]
I wanted to add: I _believe_ this issue can only cause a problem if the client and server run in the same process, and panics are swallowed without cleanup. It's still worth fixing, but I do not expect it to come up that often in ordinary workloads.

**Comment by creachadair** [comment]
When you are able, can you please patch #42 and let me know whether it addresses your cases?

**Comment by Lachann** [comment]
I'm definitely seeing this when running client and server in distinct processes - see attached files. 
[pushcall.zip](https://github.com/creachadair/jrpc2/files/5954188/pushcall.zip)

The #42 patch does fix it to an extent, in that a panic in the callback handler terminates the `PushCall` on server with an appropriate error. However, if the client process is interrupted in other ways while `PushCall` is in progress (for example, via `SIGINT`), the call still hangs server-side. I've tried adding deferred cleanup client-side, to make sure the client connection is closed with `(*jrpc2.Client).Close`, but it didn't seem to help.

**Comment by creachadair** [comment]
Thank you for the additional detail. I believe I understand what is happening now.

**Comment by creachadair** [comment]
I have updated #42 to address this problem. In short: Call handlers were not being correctly unblocked during server shutdown, and now are. It fixes the problem in your example code (which was very helpful, thank you again), but can you please verify whether it works on your real workload too?

**Comment by Lachann** [comment]
It works :) Thank you!

**Comment by creachadair** [comment]
@Lachann Thank you very much for your patient and complete report. I really appreciate it. I'll go ahead and merge the fix and push a new tag.

**Comment by creachadair** [comment]
The fix is now release in tags at or after v0.11.3.
